packages/cli/src/runAndroid/adb.js

```diff
@@ -34,18 +35,45 @@ function parseDevicesResult(result: string): Array<string> {
 /**
  * Executes the commands needed to get a list of devices from ADB
  */
-function getDevices(): Array<string> {
+function getDevices(adbPath: string): Array<string> {
   try {
-    const devicesResult = execSync('adb devices');
+    const devicesResult = execSync(`${adbPath} devices`);
     return parseDevicesResult(devicesResult.toString());
   } catch (e) {
+    logger.error(e.message);
+    return [];
+  }
+}
+
+/**
+ * Gets available CPUs of devices from ADB
+ */
+function getAvailableCPUs(adbPath: string, device: string): Array<string> {
+  try {
+    const baseArgs = ['-s', device, 'shell', 'getprop'];
+
+    let cpus = execFileSync(
+      adbPath,
+      baseArgs.concat(['ro.product.cpu.abilist'])
+    ).toString();
+
+    // pre-Lollipop
+    if (!cpus || cpus.trim().length === 0) {
+      cpus = execFileSync(
+        adbPath,
+        baseArgs.concat(['ro.product.cpu.abi'])
+      ).toString();
+    }
+
+    return (cpus || '').trim().split(',');
+  } catch (e) {
+    logger.error(e.message);
```

Comment:
We shouldn't capture errors here, log and return an empty array. Let's let this error go up forward - there's a global try/catch that will print the error. 

You could throw a new error here, if you want to make a better message.

Reply:
@grabbou I don't think we need to log errors here. There is no meaningful information.

Reply:
Then, I think it's a question whether returning an empty array in case of an error and swallowing it is a good design decision or not.

I would assume we throw an error - is an empty array of CPUs a "good" return value? When this can happen?

Reply:
@grabbou I agree with your idea. I’m sure that the probability of this error you said is very low. However, it returns an empty array for defense logic in case of an unknown situation.

Reply:
> it returns an empty array for defense logic

I am just worried that returning an empty array may provide confusing and hard to debug results when this error actually happens.

My question is:
- is it possible for an app to have no CPU architectures available? 
- is the outer code (in our case - runAndroid) able to handle this case? 

We should focus on letting the user know something bad happened so that they can debug it.

That said, I think it would be better to just throw an error here, e.g. `throw new Error("There was an error getting an array of XXXX, consider doing Y")`

Reply:
@grabbou 
As far as I know, using `adb` can get the CPU architectures on all android devices. If an error occurs, it will be the part of running that. It may be an unknown issue. In this case, it is a matter of choosing whether to exit the program without trying to install the APK.

I chose to try to install even if I didn't know the CPU architectures.

- Even if the CPU array is empty, try to install the APK in two cases. [[code](https://github.com/ifsnow/react-native-cli/blob/fix/run-android-with-deviceId/packages/cli/src/runAndroid/runAndroid.js#L199)]
  - case 1) app-universal-debug.apk
  - case 2) app-debug.apk (`enableSeparateBuildPerCPUArchitecture = false`)
- If an installable APK is not found, throw an error. [[code](https://github.com/ifsnow/react-native-cli/blob/fix/run-android-with-deviceId/packages/cli/src/runAndroid/runAndroid.js#L212)]

It is very unlikely that this step will happen, so you don't have to worry about it. :)

Reply:
Okay, thank you for this answer and explaining it. Makes sense :)